drivers/can/can_mcan.c

```diff
@@ -752,13 +756,22 @@ int can_mcan_send(const struct can_mcan_config *cfg,
 	CACHE_CLEAN(&msg_ram->tx_buffer[put_idx].hdr, sizeof(tx_hdr));
 	CACHE_CLEAN(&msg_ram->tx_buffer[put_idx].data_32, ROUND_UP(data_length, 4));
 
-	data->tx_fin_cb[put_idx] = callback;
-	data->tx_fin_cb_arg[put_idx] = user_data;
+	if (!cfg->one_shot) {
+		data->tx_fin_cb[put_idx] = callback;
+		data->tx_fin_cb_arg[put_idx] = user_data;
+	}
 
 	can->txbar = (1U << put_idx);
 
 	k_mutex_unlock(&data->tx_mtx);
 
+	if (cfg->one_shot) {
+		if (callback != NULL) {
+			callback(0, user_data);
+		}
+		return 0;
+	}
+
```

Comment:
Callbacks should be called regardless of "one-shot" mode being enabled or not, they should not require special handling here?

Reply:
I did this because I thought that the callback wouldn't fire if there's no ACK on bus, and in one-shot mode we don't rely on that.

Reply:
To elaborate, the callback is normally fired from here https://github.com/zephyrproject-rtos/zephyr/blob/224792b2c897347d94e358a1a6f83d90d62c1719/drivers/can/can_mcan.c#L489, but that won't happen if there's no transmit complete event.

Reply:
Doing some more testing on this, those events definitely don't happen in the case that DAR is set and the device is not connected to the bus.

The tx buffer isn't cleared either, so if you attempt to send repeatedly, you'll get a 0 result until you fill up the tx buffers, and then you'll get -EAGAIN.

Reply:
@henrikbrixandersen Thanks for your comments here and on Discord! After looking at this further, I agree that these changes to callback handling aren't right.

If the frame is sent successfully with DAR set, the normal callback mechanism works.

So, I think with these callback changes removed, this PR handles the successful path properly. However, it doesn't handle cases where the frame isn't sent successfully. Here's the relevant section in the reference manual (RM0433 Rev 7):

> **Frame transmission in disabled automatic retransmission (DAR) mode**
> In DAR mode all transmissions are automatically canceled after they started on the CAN bus. A Tx buffer Tx request pending bit FDCAN_TXBRP.TRPx is reset after successful transmission, when a transmission has not yet been started at the point of cancellation, has been aborted due to lost arbitration, or when an error occurred during frame transmission.
> * Successful transmission:
>   - Corresponding Tx buffer transmission occurred bit FDCAN_TXBTO.TOx set
>   - Corresponding Tx buffer cancellation finished bit FDCAN_TXBCF.CFx not set
> * Successful transmission in spite of cancellation:
>   - Corresponding Tx buffer transmission occurred bit FDCAN_TXBTO.TOx set
>   - Corresponding Tx buffer cancellation finished bit FDCAN_TXBCF.CFx set
> * Arbitration loss or frame transmission disturbed:
>   - Corresponding Tx buffer transmission occurred bit FDCAN_TXBTO.TOx not set
>   - Corresponding Tx buffer cancellation finished bit FDCAN_TXBCF.CFx set
>
> In case of a successful frame transmission, and if storage of Tx events is enabled, a Tx event FIFO element is written with event type ET = 10 (transmission in spite of cancellation).

In order to fulfill the `can_send()` [return values](https://docs.zephyrproject.org/apidoc/latest/group__can__interface.html#ga1bf5353bf15ffc252726e71ace9f5372) we should look at `FDCAN_TXBCF.CFx` and `FDCAN_TXBTO.TOx`. I tried to enable the TCF (transmission cancellation finished) interrupt, since that seemed like a good place to check those registers, but it doesn't seem like that interrupt is fired when DAR auto-cancels. Can anyone with more knowledge comment on that?

Reply:
@henrikbrixandersen I'm going to resolve this conversation and move my question about the unhappy cases with DAR to a top-level comment.